piq/iw_ssim.py

```diff
@@ -0,0 +1,459 @@
+evaluation.
+
+Estimation values produced by presented implementation corresponds to MATLAB based estimations [2].
+
+References:
+  [1] Wang, Zhou, and Qiang Li.
+    "Information content weighting for perceptual image quality assessment."
+    IEEE Transactions on image processing 20.5 (2011): 1185-1198.
+    https://ece.uwaterloo.ca/~z70wang/publications/IWSSIM.pdf
+
+  [2] https://ece.uwaterloo.ca/~z70wang/research/iwssim/
+"""
+
+import torch
+from torch.nn.modules.loss import _Loss
+import torch.nn.functional as F
+from piq.utils import _validate_input, _reduce, _parse_version
+from piq.functional import gaussian_filter, binomial_filter1d, average_filter2d, rgb2yiq
+from typing import Union, Optional, Tuple
+import math
+
+
+def information_weighted_ssim(x: torch.Tensor, y: torch.Tensor, data_range: Union[int, float] = 1.,
+                              kernel_size: int = 11, kernel_sigma: float = 1.5, k1: float = 0.01, k2: float = 0.03,
+                              parent: bool = True, blk_size: int = 3, sigma_nsq: float = 0.4,
+                              scale_weights: Optional[torch.Tensor] = None,
+                              reduction: str = 'mean') -> torch.Tensor:
+    r"""Interface of Information Content Weighted Structural Similarity (IW-SSIM) index.
+    Inputs supposed to be in range ``[0, data_range]``.
+
+    Args:
+        x: An input tensor. Shape :math:`(N, C, H, W)`.
+        y: A target tensor. Shape :math:`(N, C, H, W)`.
+        data_range: Maximum value range of images (usually 1.0 or 255).
+        kernel_size: The side-length of the sliding window used in comparison. Must be an odd value.
+        kernel_sigma: Sigma of normal distribution for sliding window used in comparison.
+        k1: Algorithm parameter, K1 (small constant).
+        k2: Algorithm parameter, K2 (small constant).
+            Try a larger K2 constant (e.g. 0.4) if you get a negative or NaN results.
+        parent: Flag to control dependency on previous layer of pyramid.
+        blk_size: The side-length of the sliding window used in comparison for information content.
+        sigma_nsq: Parameter of visual distortion model.
+        scale_weights: Weights for scaling.
+        reduction: Specifies the reduction type:
+            ``'none'`` | ``'mean'`` | ``'sum'``. Default:``'mean'``
+
+    Returns:
+        Value of Information Content Weighted Structural Similarity (IW-SSIM) index.
+
+    References:
+        Wang, Zhou, and Qiang Li..
+        Information content weighting for perceptual image quality assessment.
+        IEEE Transactions on image processing 20.5 (2011): 1185-1198.
+        https://ece.uwaterloo.ca/~z70wang/publications/IWSSIM.pdf DOI:`10.1109/TIP.2010.2092435`
+
+    Note:
+        Lack of content in the image could lead to NaN values.
+    """
+    assert kernel_size % 2 == 1, f'Kernel size must be odd, got [{kernel_size}]'
```

Comment:
Shall we also validate here that `parent` values are valid (in some `(0, min_size] range)?

Reply:
`parent` is a `bool` flag. Do you suggest to validate its type?